### ISSUE TITLE
Non-arabic scripts (Japanese) do not auto-populate slug

### ISSUE THREAD

**Issue by minikomi** [torchbox/wagtail#1443]
This was partially addressed in #8, but I'm evaluating wagtail for creating pages for Japan, and unfortunately django's URLify (understandably) doesn't stretch as far to include Japanese. I'm wondering what the best plan of attack is. Some ideas:

* If the title is non-blank, but urlify returns a blank string, use something else?
  * Date? 
  * Post ID
  * Configurable from settings or model
* Just show a warning you need to set the slug too
* Show the current slug in the main panel under the title - perhaps clicking on it could bring up the second tab and focus the slug input

**Comment by gasman** [comment]
I've previously used the [unidecode](https://pypi.python.org/pypi/Unidecode) library to produce passable ASCII representations of non-Latin text - maybe that could be used here. (I don't know how well it handles kanji in general, though - I only used it on place names. And maybe this is just pushing the problem further down the road, and eventually we'll encounter another language that unidecode can't handle...)

I guess there's no fundamental reason why slugs have to be ASCII-only - most browsers now display unicode characters in URLs quite happily - so perhaps the answer is simply to use something more permissive than django's urlify. (What do Japanese news/blog sites usually do for URLs - native script, romanized text, or avoid the issue by just using IDs?)

**Comment by minikomi** [comment]
I've seen all 3 used before. To me, romanize versions of th Japanese titles
are the worst option - I can't see it having any SEO benefit whatsoever and
doesn't really make the URLs any more typable.

To be honest, this is something I feel might be better handled in the model
rather than on the front end. There's lots of cases where a datestamp would
be good enough or perhaps just the post's id.

Eg - if the slug is blank run generate_slug method, if that returns blank
string or none raise validation error.

Actually, using the routableurls mixin and an index page, I can see cases
where the slug is never used and the url comes from eg. The post date / ID
and lookup is ID based.

**Comment by kaedroho** [comment]
A commit recently landed in Django adding support for unicode characters in ``SlugField``: https://github.com/django/django/commit/f8cc464452f495fce2a3d6f7494396c8f798a1e6

**Comment by hanpama** [comment]
```
In [2]: from demo.models import BlogPage

In [3]: BlogPage.objects.all()
Out[3]: [<BlogPage: 블로그를 열었습니다>, <BlogPage: 두 번째 블로그입니다>, <BlogPage: 세번째 블로그입니다.>]

In [4]: BlogPage.objects.all()[0]
Out[4]: <BlogPage: 블로그를 열었습니다>

In [5]: BlogPage.objects.all()[0].slug
Out[5]: 'hello'

In [6]: model_obj = BlogPage.objects.all()[0]

In [7]: model_obj.slug = "한글한글"

In [8]: model_obj.save()

In [9]: model_obj.slug
Out[9]: '한글한글'
```
<img width="753" alt="2015-10-15 4 25 32" src="https://cloud.githubusercontent.com/assets/7418482/10507220/f1615d6e-7359-11e5-84b2-ab1cdb047e07.png">

I manually manipulated model instance's slug field with unicode string and saved it.
And the unicode slug is working well. (It will be better to write some tests for unicode slug)

But the problem is that when user types unicode string as a page's title in admin,
js automatically cleans non-ascii characters, so I am rewriting codes bellow to work well with unicode..

```javascript
function cleanForSlug(val, useURLify) {
    if (URLify != undefined && useURLify !== false) { // Check to be sure that URLify function exists, and that we want to use it.
        return URLify(val, val.length);
    } else { // If not just do the "replace"
        return val.replace(/\s/g, '-').replace(/[^A-Za-z0-9\-\_]/g, '').toLowerCase();
    }
}
```

https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailadmin/static_src/wagtailadmin/js/page-editor.js#L297-L324

Uhm.. do you think this approach is good?

**Comment by gasman** [comment]
@hanpama Allowing non-ASCII characters in slugs seems to be the right solution, yes. To do this, we would need to update the validation code in both Python and Javascript. In both cases, we want it to accept all characters that are considered alphanumeric in Unicode (not just A-Za-z0-9). Python has [built-in support for this](https://docs.python.org/2/library/re.html#re.UNICODE), but it seems that Javascript doesn't - we'd have to list all the valid character ranges, as in http://stackoverflow.com/a/22075070/1853523.

On the Python side, we would need to override the default validation on SlugField: https://docs.djangoproject.com/en/1.8/ref/forms/validation/#using-validators
(Your example above manages to bypass this validation because you're writing directly to the model, but the validation still happens at the form level.)

On the Javascript side, we'd need to update the cleanForSlug implementation. We can't use the URLify function from django.contrib.admin (because the current version in Django 1.8 is not unicode-aware, and we won't be dropping Django 1.8 support any time soon) - so we'll have to write our own replacement (including the extra functionality such as removal of stop-words), or bundle our own copy of the Django 1.9 version.

**Comment by hanpama** [comment]
Thank you, @gasman .
I will make some updates on this issue, and send PR soon!

**Comment by akhayyat** [comment]
When closed, will the fix to this issue allow prepopulating the slug field with unicode characters if the title is contains non-Ascii unicode characters?

**Comment by hanpama** [comment]
To allow unicode slug, we have to define a class which extends `django.db.models.SlugField` and replace page model's slug field with custom one. Django 1.9 supports unicode slug, so using its SlugField seems nice but backward compatibility matters.

Do you think it is a nice solution to make a new custom `UnicodeSlugField`? or... is there a better solution for this?

**Comment by gasman** [comment]
Fixed (on Django 1.9 - Django 1.8 falls back on the old behaviour) in #2396.
